feat(gnome-extensions): Add PK ID recipe input support

fiftydinar opened the pull request:
# Subject

PK number ID config can be used now for recipe input, in case if any duplicated extensions exist.
Module conveniently informs about this scenario & recommends the user to switch to PK ID in this case.

Mixed config in PK ID & "literal extension names" are supported for both install & uninstall entries.

Additionally not related to this PR's subject,
I improved reliability of applying missing gschema.compiled workaround.
I use UUID detetection now instead of "literal extension name".

# Status

Apart from maybe some doc updates, it's ready to merge.

Wondering If I should add additional documentation besides module.yml or not.

fiftydinar committed:
chore(gnome-extensions): Make extension URL quering more robust

If there are multiple Gnome extensions with the same name, & they are compatible with Gnome version in the base image, then build will fail.

Previously, build would always fail if there are extensions with duplicate names.

fiftydinar committed:
docs(gnome-extensions): Fix typo

fiftydinar committed:
docs(gnome-extensions): Mention github repo

fiftydinar committed:
fix(gnome-extensions): Version checking

fiftydinar committed:
chore(gnome-extensions): Make operation order better for reading

xynydev commented on modules/gnome-extensions/README.md:
```diff
@@ -55,3 +55,27 @@ How to uninstall extensions using the module:
+
+## Known issues
+
+### Multiple compatible Gnome extensions with the same name are not supported
```
One would probably still want these sorts of extensions installed, so wouldn't it make sense to recommend one to use the legacy method here? That's what I opted for with Panel Date Format. I believe the legacy method would need to be redocumented, though.

xynydev replied:
Also, one module definition currently cant mix legacy & new config types, but that does make it a bit more clumsy to use both.

fiftydinar replied:
I'll think about refactoring the code to additionally support UUID in some way, rather than advising users to use legacy config.

fiftydinar replied:
@xynydev I find out how to support specifying UUID:

We should use `extension-info` instead of `extension-query`
Thanks to looking at Extension Manager source code.

Example of newer `Panel Date Format:`  
https://extensions.gnome.org/extension-info/?pk=1462

Example of old `Panel Date Format`:  
https://extensions.gnome.org/extension-info/?pk=3465

I'll try to think on what is the best way to handle this, the transition to new module input in recipe again, since specifying UUID is the most preferred way.

fiftydinar replied:
I'm also not sure If I should ditch this PR & start working on new one, or merge this, as a band-aid workaround.

xynydev replied:
I think this PR is alright. We could support both specifying extensions with the names and with the pk/id. The error messages from this PR could then be amended to tell the user to use the extension id instead.

fiftydinar committed:
feat(gnome-extensions): Use new PK ID installation method

fiftydinar committed:
fix(gnome-extensions): Unbound variable error

fiftydinar committed:
fix(gnome-extensions): Legacy config order logic

fiftydinar committed:
Merge branch 'main' into gnome-ext-fix-same-name

fiftydinar committed:
docs(gnome-extensions): Fix spacing

fiftydinar committed:
docs(gnome-extension): Delete false-claim about better config readability

fiftydinar committed:
Merge branch 'main' into gnome-ext-fix-same-name

fiftydinar committed:
chore(gnome-extensions): Add safe check for "null" in if `jq` parsing

fiftydinar committed:
chore(gnome-extensions): Update uninstall step to use PK

This will break config for uninstalling extensions in literal names, but I bet that no one uses this feature yet since this extension is still new & fresh, so I'll just break it.

fiftydinar committed:
Merge branch 'main' into gnome-ext-fix-same-name

xynydev commented:
IMO we shouldn’t deprecate what you call “literal name config” because:
- We just added it, deprecating it / calling it legacy immediately might be confusing for users and make the system feel less stable
- It works well for extensions where the names are unique and errors out for extensions where they aren’t, keeping it shouldn’t cause additional user frustration
- It produces more readable configuration files and has kind of a wow-effect (so you just install the extensions by typing their names there? cool), whereas IDs feel technical and produce unreadable configuration that basically requires comments

fiftydinar commented:
>     * We just added it, deprecating it / calling it legacy immediately might be confusing for users and make the system feel less stable

I understand this, but it's better to fix the mistake now, rather than later.

>     * It works well for extensions where the names are unique and errors out for extensions where they aren’t, keeping it shouldn’t cause additional user frustration

It is a rare scenario, but it can happen that user's extension which he initially installed gets incompatible, while some other irrelevant extension with the same name is compatible & it gets installed without user knowing.

That's something that I'm not comfortable with.

>     * It produces more readable configuration files and has kind of a wow-effect (so you just install the extensions by typing their names there? cool), whereas IDs feel technical and produce unreadable configuration that basically requires comments

I agree with this, but see the point above where I'm uncomfortable.

Maybe that scenario realistically never happens, so we can keep both "literal-name" & "PK".

xynydev commented:
> It is a rare scenario, but it can happen that user's extension which he initially installed gets incompatible, while some other irrelevant extension with the same name is compatible & it gets installed without user knowing.
> 
> That's something that I'm not comfortable with.

Can’t the module be made to error every time there are duplicate extensions names, even when the versions are different and one is incompatible? 

The only scenario I see after that is an extension changing its name and another extension popping up with the same name as what that extension had previously, which would effectively make the user get something else than what they wanted. I find this unlikely too.

fiftydinar commented:
> Can’t the module be made to error every time there are duplicate extensions names, even when the versions are different and one is incompatible?

I can do that too, while telling the user to input PK in recipe file instead if this happens.

fiftydinar committed:
Merge branch 'main' into gnome-ext-fix-same-name

fiftydinar committed:
Revert to the current module state

fiftydinar committed:
docs(gnome-extensions): Add README details about missing gschema.compiled file

fiftydinar committed:
chore(gnome-extensions): Introduce hybrid literal-name/PK config

If 2 or more same-named extensions are recommended, then user is recommended to use PK ID instead.

fiftydinar committed:
chore(gnome-extensions): Ignore ShellCheck 2128 error, since it's harmless & intended

fiftydinar committed:
chore(gnome-extensions): Also ignore harmless SC2178 notice

fiftydinar committed:
chore(gnome-extensions): Use UUID for matching compiled.gschema workaround rather than extension name

fiftydinar committed:
docs(gnome-extensions): Add note in module.yml that PK can be used

fiftydinar commented:
@xynydev PR is now ready, please read the updated 1st post, especially "Status".
